File: src/algorithms/datastructure/MergeIntervals.java
package algorithms.datastructure; 

import java.util.*; 
import lib.Interval; 

public class MergeIntervals {
    /* Function: Given a list of intervals, merge intervals if they can and output all new of intervals. 
     * Solution: First sort them by start time. If the end of current interval is small than the next one's, 
     *     this interval can be outputed. Else, they can merge, keep the start but let end be the larger one. 
     */ 
    
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> res = new ArrayList<Interval>(); 
        if (intervals == null || intervals.size() == 0) {
            return res; 
        }
        
        Queue<Interval> heap = new PriorityQueue<Interval>(intervals.size(), IntervalComparator); 
        for (Interval itv : intervals) {
            heap.offer(itv); 
        }
        
        Interval cur = heap.poll(); 
        int st = cur.start, ed = cur.end; 
        while (!heap.isEmpty()) {
            cur = heap.poll(); 
            if (ed < cur.start) {
                res.add(new Interval(st, ed)); 
                st = cur.start; 
                ed = cur.end; 
            } else {
                ed = Math.max(ed, cur.end); 
            }
        }
        
        res.add(new Interval(st, ed)); 
        return res; 
    }
    
    private Comparator<Interval> IntervalComparator = new Comparator<Interval>() {
        public int compare(Interval i1, Interval i2) {
            return i1.start - i2.start; 
        }
    }; 
}

File: wwwroot/js/merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    if (intervals.length === 0) {
        return;
    }

    intervals.sort(function(a, b) { return a[0] - b[0]; });

    let res = [];

    let st = intervals[0][0];
    let ed = intervals[0][1];
    for (let i = 1; i < intervals.length; i++) {
        let cur = intervals[i];
        if (cur[0] > ed) {
            console.log(st, ed);
            res.push([st, ed]);
            st = cur[0];
            ed = cur[1];
        } else {
            ed = Math.max(ed, cur[1]);
        }
    }

    res.push([st, ed]);
    return res;
};

document.getElementById("merge-intervals").innerHTML =
    merge([[1,3],[2,6],[8,10],[15,18]]);
